File: Windows/Prosary/Models/BasicPrayersOrder.cs
using Windows.Storage;

namespace Prosary.Models;

/// <summary>The user's personal ordering of the basic-prayers list (Erez, 2026-08-08) — the
/// <see cref="HomeOrder"/> pattern on a fixed catalog: a persisted list of prayer ids, catalog
/// order for ids it does not name (so a prayer added in an update appears after the ordered
/// ones), an empty list meaning pure catalog order. Mirrors iOS/Android.</summary>
public static class BasicPrayersOrder
{
    private const string Key = "basicPrayerOrder";

    public static IReadOnlyList<string> Saved()
    {
        try
        {
            return ApplicationData.Current.LocalSettings.Values[Key] is string csv && csv.Length > 0
                ? csv.Split('\n')
                : [];
        }
        catch
        {
            return [];
        }
    }

    public static void Save(IEnumerable<string> ids)
    {
        try
        {
            ApplicationData.Current.LocalSettings.Values[Key] = string.Join('\n', ids);
        }
        catch
        {
            // Settings I/O never breaks the list.
        }
    }

    public static void Reset()
    {
        try
        {
            ApplicationData.Current.LocalSettings.Values.Remove(Key);
        }
        catch
        {
        }
    }

    /// <summary>Stable: unknown ids keep their relative (catalog) order after the ordered.</summary>
    public static List<BasicPrayer> Apply(IEnumerable<BasicPrayer> prayers)
    {
        var order = Saved().ToList();
        var list = prayers.ToList();
        if (order.Count == 0)
        {
            return list;
        }

        return list
            .Select((prayer, index) => (prayer, index))
            .OrderBy(x => { var i = order.IndexOf(x.prayer.Id); return i < 0 ? int.MaxValue : i; })
            .ThenBy(x => x.index)
            .Select(x => x.prayer)
            .ToList();
    }
}

File: Android/app/src/main/java/com/dkaluta/prosary/models/BasicPrayersOrder.kt
package com.dkaluta.prosary.models

import android.content.Context

/** The user's personal ordering of the basic-prayers list (Erez, 2026-08-08) — the [HomeOrder]
 * pattern on a fixed catalog: a persisted list of prayer ids, catalog order for ids it does
 * not name (so a prayer added in an update appears after the ordered ones), an empty list
 * meaning pure catalog order. Mirrors iOS's BasicPrayersOrder. */
object BasicPrayersOrder {
    private const val PREFS = "basic_prayers_order"
    private const val KEY = "order"

    fun saved(context: Context): List<String> =
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .getString(KEY, null)?.split('\n')?.filter { it.isNotEmpty() } ?: emptyList()

    fun save(context: Context, ids: List<String>) {
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .edit().putString(KEY, ids.joinToString("\n")).apply()
    }

    fun reset(context: Context) {
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit().remove(KEY).apply()
    }

    /** Stable: unknown ids keep their relative (catalog) order after the ordered ones. */
    fun apply(context: Context, prayers: List<BasicPrayer>): List<BasicPrayer> {
        val order = saved(context)
        if (order.isEmpty()) return prayers
        return prayers.withIndex().sortedWith(
            compareBy({ order.indexOf(it.value.id).let { i -> if (i < 0) Int.MAX_VALUE else i } }, { it.index }),
        ).map { it.value }
    }
}
